WIP on elliptic fibrations again

HechtiDerLachs opened the pull request:
CC: @simonbrandhorst

HechtiDerLachs committed:
Squashed changes.

HechtiDerLachs committed:
Remove methods for preimages.

HechtiDerLachs committed:
Add some methods for dim and remove debug messages.

HechtiDerLachs committed:
Revert temporary workarounds for MPolyQuo.jl.

HechtiDerLachs committed:
Revert hotfixes for modules over quotient rings.

HechtiDerLachs committed:
Remove duplicate line.

HechtiDerLachs committed:
Remove deprecated code.

HechtiDerLachs committed:
Revert "Remove methods for preimages."

This reverts commit dbeb35d07150f361b72e1c60c1e1c41fdb863dc7.

HechtiDerLachs committed:
Revert changes in derived pushforwards.

HechtiDerLachs committed:
Clean up docstrings for preimages of ideals.

HechtiDerLachs committed:
Fix docstrings.

HechtiDerLachs committed:
Fix setting of attributes.

HechtiDerLachs committed:
Remove changes by only trailing whitespaces.

HechtiDerLachs committed:
Remove typo.

HechtiDerLachs committed:
Revert "Remove typo."

This reverts commit 50978be23a1a04de1124e241f92ae27dfb0216f5.

HechtiDerLachs committed:
Remove typo.

HechtiDerLachs committed:
Fix dummy method for docstring.

HechtiDerLachs committed:
Reroute computation of dimensions for localized ideals at rational points.

HechtiDerLachs committed:
Add tests.

HechtiDerLachs committed:
Fix test file for covered schemes.

HechtiDerLachs committed:
Fix docs.

HechtiDerLachs committed:
Fix doctests.

HechtiDerLachs committed:
Fix method ambiguity.

HechtiDerLachs committed:
Move docstring and export abstract type.

HechtiDerLachs committed:
Avoid factorization in simplification of multiplicative sets.

HechtiDerLachs committed:
Avoid slow multiplication in creation of pos.

HechtiDerLachs committed:
Remove method ambiguity.

HechtiDerLachs committed:
Fix tests.

HechtiDerLachs committed:
Repair creation of prime ideals from charts.

HechtiDerLachs committed:
Remove weird preimage tests.

HechtiDerLachs committed:
Use dimension of leading ideal.

HechtiDerLachs committed:
Update experimental/Schemes/IdealSheaves.jl

Co-authored-by: Simon Brandhorst <brandhorst@math.uni-sb.de>

HechtiDerLachs committed:
Remove too expensive assertions again.

HechtiDerLachs committed:
Remove debug message.

HechtiDerLachs committed:
Fix tests.

HechtiDerLachs committed:
WIP on matching of divisors.

HechtiDerLachs committed:
Remove assertion from debugging.

HechtiDerLachs committed:
Adjust signatures for polynomial maps.

Co-authored-by: Johannes Schmitt <jschmitt@posteo.eu>

HechtiDerLachs committed:
Add some type stability.

HechtiDerLachs committed:
Avoid expensive is_subset computation.

HechtiDerLachs committed:
WIP on pushing around divisors.

HechtiDerLachs committed:
Fix tests.

HechtiDerLachs committed:
Fix tests.

HechtiDerLachs committed:
WIP on pushing divisors around.

HechtiDerLachs committed:
Fix constructor.

HechtiDerLachs committed:
Fix tests.

HechtiDerLachs committed:
Customize kernel computations.

HechtiDerLachs committed:
Work on pushforward of divisors [no ci].

HechtiDerLachs committed:
Fix creation of ideal sheaves.

HechtiDerLachs committed:
Fix tests.

HechtiDerLachs committed:
Re-enable simplified mapping.

HechtiDerLachs committed:
Re-enable old simplify method for MPolyPowersOfElement.

HechtiDerLachs committed:
WIP.

HechtiDerLachs committed:
Fix bug.

HechtiDerLachs committed:
Fix tests.

HechtiDerLachs committed:
Remove some debug messages.

simonbrandhorst commented on docs/src/AlgebraicGeometry/Schemes/MorphismsOfProjectiveSchemes.md:
```diff
@@ -22,9 +22,9 @@ out.
-    morphism_of_projective_schemes(X::AbsProjectiveScheme, Y::AbsProjectiveScheme, a::Vector{<:RingElem})
+    morphism(P::AbsProjectiveScheme, Q::AbsProjectiveScheme, f::Map; check::Bool=true )
+    morphism(P::AbsProjectiveScheme, Q::AbsProjectiveScheme, f::Map, h::SchemeMor; check::Bool=true )
+    morphism(X::AbsProjectiveScheme, Y::AbsProjectiveScheme, a::Vector{<:RingElem})
```
please deprecate

simonbrandhorst commented on experimental/Schemes/IdealSheaves.jl:
```diff
@@ -544,10 +557,40 @@ end
+function is_one(I::SumIdealSheaf; covering::Covering=default_covering(scheme(I)))
+  return get_attribute!(I, :is_one) do
+    J = summands(I)
+    k = findfirst(x->x isa PrimeIdealSheafFromChart, J)
+    if k !== nothing
+      P = J[k]
+      U = original_chart(P)
+      if !is_one(cheap_sub_ideal(I, U))
```
```suggestion
    k = findfirst(x->x isa PrimeIdealSheafFromChart, J)
    if k !== nothing
      # find a chart where we can cheaply compute a summand
      P = J[k]
      U = original_chart(P)
      if !is_one(cheap_sub_ideal(I, U))
```

simonbrandhorst commented on experimental/Schemes/IdealSheaves.jl:
```diff
@@ -544,10 +557,40 @@ end
+
+    for U in keys(object_cache(I))
+      !is_one(I(U)) && return false
+    end
```
could be moved to the front

simonbrandhorst commented on experimental/Schemes/IdealSheaves.jl:
```diff
@@ -1468,21 +1514,27 @@ function produce_object(F::PrimeIdealSheafFromChart, U2::AbsAffineScheme)
   if has_ancestor(x->(x===V), U2)
-    return OOX(V, U2)(F(V))
+    result = OOX(V, U2)(F(V))
+    #@assert is_one(result) || is_prime(result)
```
could be turned into an `hassert` with an assertion level.

simonbrandhorst commented on experimental/Schemes/IdealSheaves.jl:
```diff
@@ -1611,6 +1670,178 @@ function produce_object(I::PullbackIdealSheaf, U::AbsAffineScheme)
+
+  # we are in the same ancestor tree, but on top of the defining chart
+  if has_ancestor(x->(x===U2), U)
+    iso = _flatten_open_subscheme(U, U2)
```
```suggestion
    iso = _flatten_open_subscheme(U, U2)  # an embedding U -> V \subseteq U2
```

simonbrandhorst commented on experimental/Schemes/IdealSheaves.jl:
```diff
@@ -1611,6 +1670,178 @@ function produce_object(I::PullbackIdealSheaf, U::AbsAffineScheme)
+    iso = _flatten_open_subscheme(U, U2)
+    iso_inv = inverse(iso)
+    pb_P = pullback(iso_inv)(P)
+    result = ideal(OO(U2), [g for g in OO(U2).(lifted_numerator.(gens(pb_P))) if !iszero(g)])
```
```suggestion
    pb_P = pullback(iso_inv)(P)
    # avoids a saturation by discarding denominators but only produces a subideal
    result = ideal(OO(U2), [g for g in OO(U2).(lifted_numerator.(gens(pb_P))) if !iszero(g)])
```

simonbrandhorst commented on experimental/Schemes/IdealSheaves.jl:
```diff
@@ -1611,6 +1670,178 @@ function produce_object(I::PullbackIdealSheaf, U::AbsAffineScheme)
+  count = 0
+  for W in fat
+    @show count = count + 1
+    @show complexity(W)
```
```suggestion
```

simonbrandhorst commented on experimental/Schemes/MorphismFromRationalFunctions.jl:
```diff
@@ -374,29 +315,29 @@ function cheap_realization(Phi::MorphismFromRationalFunctions, U::AbsAffineSchem
   end
   img_gens_frac = realization_preview(Phi, U, V)
   # Try to cancel the fractions heuristically
```
```suggestion
  # Try to cancel the fractions heuristically; turns out it was too expensive in some applications due to slow divide
```

simonbrandhorst commented on experimental/Schemes/MorphismFromRationalFunctions.jl:
```diff
@@ -758,6 +673,81 @@ function _try_pullback_cheap(phi::MorphismFromRationalFunctions, I::AbsIdealShea
+  end
+
+  return nothing
+end
+
+function _pullback(phi::MorphismFromRationalFunctions, I::PrimeIdealSheafFromChart)
+  V0 = original_chart(I)
+  X = domain(phi)
+  Y = codomain(phi)
+  V = affine_charts(Y)
+  U = affine_charts(X)
+
+  cod_patches = AbsAffineScheme[V0]
+  cod_patches = vcat(cod_patches, [U for U in keys(object_cache(I)) if any(x->x===U, affine_charts(Y))])
+  cod_patches = vcat(cod_patches, [U for U in affine_charts(Y) if !any(x->x===U, cod_patches)])
+  for U0 in U
+    I_undef = ideal(OO(U0), elem_type(OO(U0))[])
+    random_realizations = IdDict{AbsAffineScheme, AbsAffineSchemeMor}()
+    for V0 in cod_patches
+      psi = random_realization(phi, U0, V0)
+      I_undef = I_undef + ideal(OO(U0), complement_equation(domain(psi)))
+      random_realizations[V0] = psi
+      if isone(I_undef)
+        break
+      end
+    end
+    if isone(I_undef)
+      J = ideal(OO(U0), elem_type(OO(U0))[])
+      for (V0, psi) in random_realizations
+        J_loc = pullback(psi, saturated_ideal(I(V0)))
+        J = J + ideal(OO(U0), lifted_numerator.(gens(J_loc)))
+      end
+      return IdealSheaf(X, U0, gens(J))
+    else
+      continue
+    end
+  end
+  error("pullback did not succeed")
+end
+
```
Is this in use? Should it stay? Can it go?

simonbrandhorst commented on experimental/Schemes/MorphismFromRationalFunctions.jl:
```diff
@@ -844,3 +833,306 @@ end
+  return morphism_from_rational_functions(X, Z, U, codomain_chart(g), imgs_U; check=false)
+end
+
+function ideal_sheaf_of_image(phi::MorphismFromRationalFunctions)
```
to be evaluated

simonbrandhorst commented on experimental/Schemes/Sheaves.jl:
```diff
@@ -574,7 +574,8 @@ function restrict(
   )
-  parent(a) === OO(ambient_scheme(U)) || return OO(U)(lift(a))
+  #parent(a) === OO(ambient_scheme(U)) || return OO(U)(lift(a))
+  parent(a) === OO(ambient_scheme(U)) || return convert(OO(U), fraction(a))
   return OO(U)(a, check=check)
```
looks strange, please check

simonbrandhorst commented on experimental/Schemes/Types.jl:
```diff
@@ -445,3 +447,121 @@ end
+@doc raw"""
+    MorphismFromRationalFunctions{DomainType<:AbsCoveredScheme, CodomainType<:AbsCoveredScheme} 
+
+A lazy type for a morphism ``φ : X → Y`` of `AbsCoveredScheme`s which is given 
```
```suggestion
A lazy type for a dominant morphism ``φ : X → Y`` of `AbsCoveredScheme`s which is given 
```

simonbrandhorst commented on experimental/Schemes/Types.jl:
```diff
@@ -445,3 +447,121 @@ end
+    )
+    @check is_irreducible(X) "domain must be irreducible"
+    @check is_irreducible(Y) "codomain must be irreducible"
+    #_find_chart(U, default_covering(X)) !== nothing || error("patch not found in domain")
```
```suggestion
    @check is_irreducible(Y) "codomain must be irreducible"
    @check dim(Y) <= dim(X) "cannot be dominant"
    #_find_chart(U, default_covering(X)) !== nothing || error("patch not found in domain")
```

simonbrandhorst commented on experimental/Schemes/WeilDivisor.jl:
```diff
@@ -386,7 +386,7 @@ function intersect(D::AbsWeilDivisor, E::AbsWeilDivisor;
       else
         I = c1 + c2
-        if dim(I) > 0
+        if !has_dimension_leq_zero(I)
```
```suggestion
        if !has_dimension_leq_zero(I) # potentially faster for localized ideals
```

simonbrandhorst commented on experimental/Schemes/WeilDivisor.jl:
```diff
@@ -401,6 +401,73 @@ function intersect(D::AbsWeilDivisor, E::AbsWeilDivisor;
   return result
 end
 
+@attr function has_dimension_leq_zero(I::Ideal)
```
could be documented

simonbrandhorst commented on experimental/Schemes/WeilDivisor.jl:
```diff
@@ -432,7 +519,10 @@ function colength(I::AbsIdealSheaf; covering::Covering=default_covering(scheme(I
         end
         J = J + ideal(OO(U), g)
-        isone(J) && break
+        if isone(J)
+          push!(patches_done, U)
+          continue
+        end
```
```suggestion
        isone(J) && break
```

simonbrandhorst commented on experimental/Schemes/WeilDivisor.jl:
```diff
@@ -414,11 +481,31 @@ end
+    if has_decomposition_info(covering)
+      h = decomposition_info(covering)[U]
+      if isone(J_cheap + ideal(OO(U), elem_type(OO(U))[OO(U)(a) for a in h])) # R(a) is not type stable for R::MPolyQuoRing
+        push!(patches_done, U)
```
```suggestion
        # push!(patches_done, U)
```

simonbrandhorst commented on experimental/Schemes/critical_locus.jl:
```diff
@@ -0,0 +1,52 @@
+function ideal_sheaf_of_critical_locus(phi::AbsCoveredSchemeMorphism)
```
Please document

Add a method
`critical_locus(phi::AbsCoveredSchemeMorphism) -> CoveredScheme, Morphism` 
?

simonbrandhorst replied:
do we need the codomain to be smooth? check it?

simonbrandhorst commented on experimental/Schemes/elliptic_surface.jl:
```diff
@@ -653,6 +650,16 @@ function weierstrass_contraction(X::EllipticSurface)
+    I_sing_new = ideal_sheaf_of_singular_locus(Y1)
+    I_sing_new = pushforward(inc_Y1, I_sing_new) + ideal_sheaf(E1) # new components only along the exc. set
+    @vprint :EllipticSurface 2 "decomposing singular locus\n"
+    I_sing_X0 = vcat(I_sing_X0, maximal_associated_points(I_sing_new))
```
```suggestion
    I_sing_X0 = vcat(maximal_associated_points(I_sing_new), I_sing_X0)
```

simonbrandhorst commented on experimental/Schemes/elliptic_surface.jl:
```diff
@@ -838,6 +845,8 @@ function standardize_fiber(S::EllipticSurface, f::Vector{<:WeilDivisor})
     for j in 1:i-1
       @vprint :EllipticSurface 4 "$(j) "
       # we know the intersections are 0 or 1
+      G[i, j] = G[j, i] = intersect(f[i], f[j])
+      continue
```
needs a proof that this is just as fast

simonbrandhorst replied:
do you have proof? ;-)

simonbrandhorst commented on experimental/Schemes/elliptic_surface.jl:
```diff
@@ -1940,3 +1949,275 @@ function _compute_mwl_basis(X::EllipticSurface, mwl_gens::Vector{<:EllipticCurve
+  R = ambient_coordinate_ring(U)
+  FR = fraction_field(R)
+  return morphism_from_rational_functions(X, B, U, V, [FR(gen(R, 3))])
+end
```
Let us call this `fibration`. Should be a field of `X`.  
This should be a morphism. No need to get it from rational functions.
use `weierstrass_contraction` and compose it. 
We can give the weierstrass_model a 3rd output as the projection to `P^1`

simonbrandhorst commented on experimental/Schemes/elliptic_surface.jl:
```diff
@@ -1940,3 +1949,275 @@ function _compute_mwl_basis(X::EllipticSurface, mwl_gens::Vector{<:EllipticCurve
+  return morphism_from_rational_functions(X, B, U, V, [FR(gen(R, 3))])
+end
+
+function _pushforward_lattice_along_isomorphism(step::MorphismFromRationalFunctions{<:EllipticSurface, <:EllipticSurface})
```
TODO comment

simonbrandhorst commented on experimental/Schemes/elliptic_surface.jl:
```diff
@@ -1940,3 +1949,275 @@ function _compute_mwl_basis(X::EllipticSurface, mwl_gens::Vector{<:EllipticCurve
+  img_gens2 = Fkkt.(img_gens)
+  # TODO: Cache?
+  return morphism_from_rational_functions(B, X, V, U, img_gens2, check=false)
+end
```
```suggestion
```
Does not work.

simonbrandhorst commented on src/AlgebraicGeometry/Schemes/AffineSchemes/Morphisms/Constructors.jl:
```diff
@@ -213,6 +212,7 @@ function compose(f::AbsAffineSchemeMor, g::AbsAffineSchemeMor)
 end
 
 
+#=
```
Seems to be duplicated from `src/AlgebraicGeometry/Schemes/AffineSchemes/Morphisms/Methods.jl`
double check and remove.

simonbrandhorst commented on src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Morphisms/Methods.jl:
```diff
@@ -72,22 +159,24 @@ with default covering
     end
   end
-  # We skip the gluings for the time being.
+  # We skip the glueings for the time being.
```
```suggestion
  # We skip the gluings for the time being.
```

simonbrandhorst commented on src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Morphisms/Methods.jl:
```diff
@@ -72,22 +159,24 @@ with default covering
-  CC = Covering(collect(keys(mor_dict)), IdDict{Tuple{AbsAffineScheme, AbsAffineScheme}, AbsGluing}())
-  inherit_gluings!(CC, default_covering(X))
+  CC = Covering(collect(keys(mor_dict)), IdDict{Tuple{AbsAffineScheme, AbsAffineSchemeMor}, AbsGlueing}())
+  inherit_glueings!(CC, default_covering(X))
```
```suggestion
  CC = Covering(collect(keys(mor_dict)), IdDict{Tuple{AbsAffineScheme, AbsAffineSchemeMor}, AbsGluing}())
  inherit_gluings!(CC, default_covering(X))
```
faulty merge

simonbrandhorst commented on src/Modules/ModulesGraded.jl:
```diff
@@ -1498,7 +1498,7 @@ function Base.show(io::IO, b::BettiTable)
     return
   end
-  
+
```
checkout this file just a change of blank lines.

simonbrandhorst commented on src/Modules/UngradedModules/FreeModuleHom.jl:
```diff
@@ -401,7 +401,6 @@ function hom(F::FreeMod, G::FreeMod)
   return GH, to_hom_map
 end
 
-
```
same

simonbrandhorst commented on src/Rings/MPolyMap/AffineAlgebras.jl:
```diff
@@ -29,7 +29,8 @@ affine_algebra_morphism_type(R::S, U::T) where {S <: Ring, T} = affine_algebra_m
 
-@attr Any function _singular_algebra_morphism(f::MPolyAnyMap)
+@attr Any function _singular_algebra_morphism(f::MPolyAnyMap{<:MPolyRing, <:Union{MPolyRing, MPolyQuoRing}, Nothing})
+  @assert coefficient_ring(domain(f)) === coefficient_ring(codomain(f))
```
```suggestion
  @assert coefficient_ring(domain(f)) === coefficient_ring(codomain(f))  "singular does not handle coefficient maps"
```

simonbrandhorst commented on src/Rings/MPolyMap/AffineAlgebras.jl:
```diff
@@ -212,11 +213,21 @@ function has_preimage_with_preimage(F::AffAlgHom, f::Union{MPolyRingElem, MPolyQ
+  error("not implemented")
+end
+
+function preimage(
```
Did a non-field work before? Such as `ZZ`? Please check / add a test so that we can spot regressions.

simonbrandhorst commented on src/Rings/MPolyMap/AffineAlgebras.jl:
```diff
@@ -227,6 +238,19 @@ function preimage(f::AffAlgHom, I::Union{MPolyIdeal, MPolyQuoIdeal})
   return ideal(D, D.(gens(prIx)))
 end
 
+function preimage(
```
please add a test for this fix.

simonbrandhorst commented on src/Rings/mpoly-ideals.jl:
```diff
@@ -591,6 +591,36 @@ Ideal generated by
+    R = base_ring(I)
+    J = ideal(R, zero(R))
+    if factor_generators
+      # In practice this will often lead to significant speedup due to reduction of degrees.
```
```suggestion
      # In practice this will often lead to significant speedup due to reduction of degrees.
      # TODO:  is the following faster? radical(ab,c) = intersect(radical(a,c),radical(b,c)) 
```

simonbrandhorst commented on src/Rings/mpoly-localizations.jl:
```diff
@@ -144,26 +144,67 @@ end
 ### simplification. 
 # Replaces each element d by its list of square free prime divisors.
 function simplify!(S::MPolyPowersOfElement)
```
These changes to `simplify!` should be reverted.

simonbrandhorst commented on src/Rings/mpoly-localizations.jl:
```diff
@@ -3018,14 +3059,31 @@ true
 end
 
-@attr function dim(I::MPolyLocalizedIdeal)
+@attr function dim(I::MPolyLocalizedIdeal{RT}) where {RT<:MPolyLocRing{<:Field, <:FieldElem, <:MPolyRing, <:MPolyRingElem, <:MPolyPowersOfElement}}
```
```suggestion
@attr function dim(I::MPolyLocalizedIdeal{RT}) where {RT<:MPolyLocRing{<:Ring, <:RingElem, <:MPolyRing, <:MPolyRingElem, <:MPolyPowersOfElement}}
```

simonbrandhorst commented on src/Rings/mpoly-localizations.jl:
```diff
@@ -3018,14 +3059,31 @@ true
   return dim(saturated_ideal(I))
 end
 
+@attr function dim(I::MPolyLocalizedIdeal{RT}) where {RT<:MPolyLocRing{<:Field, <:FieldElem, <:MPolyRing, <:MPolyRingElem, <:MPolyComplementOfPrimeIdeal}}
```
```suggestion
@attr function dim(I::MPolyLocalizedIdeal{RT}) where {RT<:MPolyLocRing{<:Ring, <:RingElem, <:MPolyRing, <:MPolyRingElem, <:MPolyComplementOfPrimeIdeal}}
```

simonbrandhorst commented on src/Rings/mpolyquo-localizations.jl:
```diff
@@ -668,7 +668,19 @@ end
 function inv(f::MPolyQuoLocRingElem{BRT, BRET, RT, RET, MPolyPowersOfElement{BRT, BRET, RT, RET}}) where {BRT, BRET, RT, RET}
   isone(f) && return f
+  
+  if length(terms(lifted_numerator(f))) > 10000
```
```suggestion
  if length(lifted_numerator(f)) > 10000
```

simonbrandhorst replied:
saves us from many allocations

simonbrandhorst commented on src/Rings/mpolyquo-localizations.jl:
```diff
@@ -1296,6 +1324,68 @@ function _as_affine_algebra(
   return id
 end
 
+@attr MPolyIdeal function kernel(
+    f::MPolyAnyMap{<:MPolyRing, 
+                   <:MPolyQuoLocRing{<:Any, <:Any, <:Any, <:Any,
+                                     <:MPolyPowersOfElement}})
```
this needs some benchmarking and/or an algorithm keyword.

simonbrandhorst commented on src/Rings/mpolyquo-localizations.jl:
```diff
@@ -1915,9 +2006,14 @@ end
 ### Some auxiliary functions
 
 @attr MPolyQuoLocalizedIdeal function radical(I::MPolyQuoLocalizedIdeal)
```
```suggestion
@attr T function radical(I::T) where {T<:MPolyQuoLocalizedIdeal}
```

simonbrandhorst commented on src/Rings/primary_decomposition_helpers.jl:
```diff
@@ -175,6 +181,7 @@ end
   return result
 end
 
+#=
```
check and delete if appropriate

simonbrandhorst commented on src/exports.jl:
```diff
@@ -999,7 +1000,6 @@ export monomials_of_degree
 export mori_cone
 export morphism
 export morphism_from_cox_variety
-export morphism_of_projective_schemes
```
remember the deprecation

simonbrandhorst commented on test/AlgebraicGeometry/Schemes/AffineRationalPoint.jl:
```diff
@@ -57,7 +57,7 @@
   pk = A2k([1,1])
   pL = A2L(pk) # conversion
-  @test_throws ErrorException pk == pL
+  # @test_throws ErrorException pk == pL # Does not throw anymore. Why?
```
check if still the case

simonbrandhorst commented on experimental/Schemes/IdealSheaves.jl:
```diff
@@ -790,7 +833,10 @@ function maximal_associated_points(I::AbsIdealSheaf; covering=default_covering(s
       isone(J) && continue
     end
     !is_one(I(U)) || continue                        ## supp(I) might not meet all components
+    dim(I(U)) == dim(I) || continue                  ## components of lesser dimension can be discarded
```
```suggestion
```
wrong

simonbrandhorst commented on experimental/Schemes/IdealSheaves.jl:
```diff
@@ -790,7 +833,10 @@ function maximal_associated_points(I::AbsIdealSheaf; covering=default_covering(s
+    dim(I(U)) == dim(I) || continue                  ## components of lesser dimension can be discarded
     components_here = minimal_primes(I(U))
+    #@assert all(p->is_prime(p), components_here)
+    @assert all(p->dim(p) == dim(I), components_here) "not all minimal primes have the correct dimension"
```
```suggestion
```

simonbrandhorst commented on test/AlgebraicGeometry/Schemes/transforms.jl:
```diff
@@ -71,13 +71,13 @@ end
   dimsL2 = [dim(a) for a in L2]
   @test length(L1) == 3
-  @test length(L2) == 2
+  @test length(L2) == 1
```
```suggestion
  @test length(L2) == 2
```

simonbrandhorst commented on test/AlgebraicGeometry/Schemes/transforms.jl:
```diff
@@ -71,13 +71,13 @@ end
   @test length(onesinL1) == 1
   @test length(zerosinL1) == 2
-  @test length(zerosinL2) == 1
+  @test length(zerosinL2) == 0
```
```suggestion
  @test length(zerosinL2) == 1
```

simonbrandhorst commented on test/Rings/mpoly-localizations.jl:
```diff
@@ -419,8 +419,8 @@ end
   @test isone(first(minimal_generating_set(phi1(J))))
 
   small_gens = small_generating_set(phi2(J))
```
add the algorithm keyword and test both

HechtiDerLachs committed:
Fix doctests.

HechtiDerLachs committed:
Fix method for preimages of ideals.

HechtiDerLachs committed:
Fix doctests again.

HechtiDerLachs committed:
More comments, explanations, and minor changes.

Co-authored-by: Simon Brandhorst <brandhorst@math.uni-sb.de>

HechtiDerLachs committed:
Change maximal associated points back to the old version.

Co-authored-by: Simon Brandhorst <brandhorst@math.uni-sb.de>

HechtiDerLachs commented on experimental/Schemes/IdealSheaves.jl:
```diff
@@ -307,6 +307,19 @@ function simplify!(I::IdealSheaf, cov::Covering=default_covering(space(I)))
+              )
+  return IdealSheaf(scheme(I), id_dict; check=false)
+end
+
```
```suggestion
```

HechtiDerLachs committed:
Remove deprecated simplify method for ideal sheaves.

HechtiDerLachs committed:
Rework the fibration projections.

HechtiDerLachs committed:
Remove debug messages.

HechtiDerLachs committed:
Repair doctests.

HechtiDerLachs committed:
Fix doctests.

HechtiDerLachs committed:
Fix doctests.

HechtiDerLachs committed:
Rework simplify_light for powers of elements and move types to an extra file.

HechtiDerLachs committed:
Switch to use of simplify_light for hypersurface complements.

HechtiDerLachs committed:
Some cleanup.

HechtiDerLachs committed:
Deprecate morphism_of_projective_schemes.

HechtiDerLachs committed:
Add some hasserts.

HechtiDerLachs committed:
Introduce keyword for small_generating_set.

HechtiDerLachs commented:
The PR seems to have become a bit bigger, because I moved the type declarations out of `src/Rings/mpoly-localizations.jl`. But that is just code migration and nothing has changed about its internals. 

Other than that, I think I addressed all relevant issues raised by the review.

HechtiDerLachs committed:
Fix tests.

simonbrandhorst commented:
Awsome, thanks a long for this huge piece of work! I wonder if we can soon enable some of the tests that were previously far too slow to ever run.
